Asgard/Communications/Classes/CbusEventManager.cs

```diff
@@ -0,0 +1,246 @@
+        }
+
+        #endregion
+
+        #region Methods
+
+        public bool? GetState(ushort nodeNumber, ushort eventNumber)
+        {
+            var key =
+                this.cbusEventStates.Keys
+                    .Where(n => !n.IsShort)
+                    .Where(n => n.NodeNumber == nodeNumber)
+                    .Where(n => n.EventNumber == eventNumber)
+                    .FirstOrDefault();
+            if (key is null) return null;
+            return this.cbusEventStates[key];
+        }
+
+        public bool? GetState(ushort eventNumber)
+        {
+            var key =
+                this.cbusEventStates.Keys
+                    .Where(n => n.IsShort)
+                    .Where(n => n.EventNumber == eventNumber)
+                    .FirstOrDefault();
+            if (key is null) return null;
+            return this.cbusEventStates[key];
+        }
+
+        public void RegisterCbusEvent<T>(ushort nodeNumber, ushort eventNumber, CbusEventCallback callback)
+            where T : class, ICbusAccessoryEvent
+        {
+            var key = CbusEventKey.Create<T>(nodeNumber, eventNumber);
+            this.cbusEventCallbacks[key] = callback;
+        }
+
+        #endregion
+
+        #region Support routines
+
+        private CbusEventKey? FindKey(ICbusAccessoryEvent cbusAccessoryEvent)
+        {
+            var result =
+                this.cbusEventCallbacks.Keys
+                    .Where(n => cbusAccessoryEvent.IsShortEvent == n.IsShort)
+                    .Where(n => n.IsShort || n.NodeNumber == (cbusAccessoryEvent as IHasNodeNumber)?.NodeNumber)
+                    .Where(n => (cbusAccessoryEvent.IsLongEvent && n.EventNumber == (cbusAccessoryEvent as IHasEventNumber)?.EventNumber) ||
+                                (cbusAccessoryEvent.IsShortEvent && n.EventNumber == (cbusAccessoryEvent as IHasDeviceNumber)?.DeviceNumber))
+                    .FirstOrDefault();
+            return result;
+        }
+
+        private CbusEventCallback? GetRegisteredCallback(ICbusAccessoryEvent cbusAccessoryEvent)
+        {
+            var key = FindKey(cbusAccessoryEvent);
+            if (key is null) return null;
+            return this.cbusEventCallbacks[key];
+        }
+
+        private void ProcessQueue()
+        {
+            while (!this.CancellationToken.IsCancellationRequested)
+            {
+                while (!this.cbusEventExceptionQueue.IsEmpty)
+                {
+                    this.cbusEventExceptionQueue.TryDequeue(out var result);
+                }
+
+                this.cbusEventExceptionQueue.Trigger.Reset();
+                this.cbusEventExceptionQueue.Trigger.Wait(500, this.CancellationToken);
+            }
+        }
+
+        private void RunCallback(CbusEventCallback callback, ICbusAccessoryEvent cbusAccessoryEvent)
+        {
+            try
+            {
+                Task.Run(() => callback(cbusAccessoryEvent, this), this.CancellationToken);
```

Comment:
Don't think this needs an extra Task.Run - the `RunCallback` is already wrapped in another `Task.Run`, and by not observing the returned task here the exceptions would be swallowed etc - better to just call the callback directly

Reply:
Good point.